# Importing Libraries

In [1]:
import warnings
warnings.simplefilter("ignore")

import pandas as pd
import plotly.express as px
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np
import altair as alt

# 

# Importing Dataset
#### This dataset will be used for complete Data Visualisation task and  for the Question 1 (Time Series) from Machine Learning for Business.

In [8]:
df = pd.read_csv('weather.csv')
df.head(10)

datetime  tempmax  tempmin  temp  feelslikemax  feelslikemin  feelslike  \
0  2021-01-01     14.2     10.9  12.6          14.2          10.9       12.6   
1  2021-01-02     15.6      6.2  10.8          15.6           4.8       10.6   
2  2021-01-03     14.4      6.9  11.0          14.4           6.1       10.9   
3  2021-01-04     15.8     10.3  13.0          15.8          10.3       13.0   
4  2021-01-05     15.3     10.8  13.5          15.3          10.8       13.5   
5  2021-01-06     14.3      7.3  11.3          14.3           7.3       11.2   
6  2021-01-07     15.3     10.5  12.8          15.3          10.5       12.8   
7  2021-01-08     17.6     11.8  14.5          17.6          11.8       14.5   
8  2021-01-09     15.3      9.0  11.7          15.3           6.0       10.3   
9  2021-01-10      9.8      7.9   8.8           7.8           4.5        6.3   

   dew  humidity  precip  ...  solarenergy  uvindex severerisk  \
0  9.0      79.0   2.747  ...          5.5        3        NaN   
1  6.2      74.5   0.000  ...          8.8        4        NaN   
2  5.9      71.4   0.000  ...          5.9        4        NaN   
3  5.2      59.8   0.000  ...          7.6        4        NaN   
4  8.5      72.0   1.202  ...          9.1        5        NaN   
5  7.0      75.9   0.121  ...          6.9        4        NaN   
6  7.9      72.0   0.000  ...          5.2        3        NaN   
7  9.1      70.5   0.000  ...          8.4        4        NaN   
8  8.7      82.6   8.563  ...          2.6        2        NaN   
9  7.7      93.1  32.574  ...          1.3        1        NaN   

               sunrise               sunset  moonphase  \
0  2021-01-01T08:29:19  2021-01-01T17:46:34       0.58   
1  2021-01-02T08:29:24  2021-01-02T17:47:25       0.61   
2  2021-01-03T08:29:27  2021-01-03T17:48:19       0.65   
3  2021-01-04T08:29:27  2021-01-04T17:49:13       0.69   
4  2021-01-05T08:29:26  2021-01-05T17:50:09       0.72   
5  2021-01-06T08:29:22  2021-01-06T17:51:07       0.75   
6  2021-01-07T08:29:15  2021-01-07T17:52:06       0.79   
7  2021-01-08T08:29:07  2021-01-08T17:53:06       0.83   
8  2021-01-09T08:28:56  2021-01-09T17:54:08       0.86   
9  2021-01-10T08:28:42  2021-01-10T17:55:10       0.90   

               conditions                                        description  \
0  Rain, Partially cloudy  Partly cloudy throughout the day with rain in ...   
1                   Clear               Clear conditions throughout the day.   
2        Partially cloudy                  Partly cloudy throughout the day.   
3        Partially cloudy                  Partly cloudy throughout the day.   
4  Rain, Partially cloudy       Clearing in the afternoon with morning rain.   
5  Rain, Partially cloudy  Partly cloudy throughout the day with late aft...   
6        Partially cloudy                  Partly cloudy throughout the day.   
7        Partially cloudy                  Partly cloudy throughout the day.   
8  Rain, Partially cloudy        Partly cloudy throughout the day with rain.   
9          Rain, Overcast  Cloudy skies throughout the day with a chance ...   

                icon                                           stations  
0               rain  LTBA,17063099999,17119099999,17061099999,17064...  
1          clear-day  LTBA,17063099999,17119099999,17061099999,17064...  
2  partly-cloudy-day  D8508,LTBA,17063099999,17119099999,17061099999...  
3  partly-cloudy-day  LTBA,17063099999,17119099999,17061099999,17064...  
4               rain  LTBA,17063099999,17119099999,17061099999,17064...  
5               rain  D8508,LTBA,17063099999,17119099999,17061099999...  
6  partly-cloudy-day  LTBA,17063099999,17119099999,17061099999,17064...  
7  partly-cloudy-day  LTBA,17063099999,17119099999,17061099999,17022...  
8               rain  D8508,LTBA,17063099999,17119099999,17061099999...  
9               rain  D8508,LTBA,17063099999,17119099999,17061099999...  

[10 rows x 32 columns]

#### For better understanding of the dataset, I'm providing terminology explanation, to be aware what each weather parameter represents:

datetime: Date and time of the weather data.

tempmax: Maximum temperature recorded.

tempmin: Minimum temperature recorded.

temp: Current temperature.

feelslikemax: Maximum "feels like" temperature.

feelslikemin: Minimum "feels like" temperature.

feelslike: Current "feels like" temperature.

dew: Dew point temperature.

humidity: Relative humidity percentage.

precip: Precipitation amount.

precipprob: Probability of precipitation.

precipcover: Extent of precipitation coverage.

preciptype: Type of precipitation (e.g., rain, snow).

snow: Snowfall amount.

snowdepth: Snow depth on the ground.

windgust: Maximum wind gust.

windspeed: Current wind speed.

winddir: Wind direction.

sealevelpressure: Atmospheric pressure at sea level.

cloudcover: Percentage of sky covered by clouds.

visibility: Visibility range in meters.

solarradiation: Solar radiation amount.

solarenergy: Solar energy captured or available.

uvindex: UV index.

severerisk: Risk of severe weather conditions.

sunrise: Time of sunrise.

sunset: Time of sunset.

moonphase: Current phase of the moon.

conditions: Summary of weather conditions.

description: Detailed description of weather conditions.

icon: Icon representing weather conditions.

stations: Information about weather monitoring stations or locations.

In [11]:
print(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2000 entries, 0 to 1999
Data columns (total 32 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   datetime          2000 non-null   object 
 1   tempmax           2000 non-null   float64
 2   tempmin           2000 non-null   float64
 3   temp              2000 non-null   float64
 4   feelslikemax      2000 non-null   float64
 5   feelslikemin      2000 non-null   float64
 6   feelslike         2000 non-null   float64
 7   dew               2000 non-null   float64
 8   humidity          2000 non-null   float64
 9   precip            2000 non-null   float64
 10  precipprob        2000 non-null   float64
 11  precipcover       2000 non-null   float64
 12  preciptype        997 non-null    object 
 13  snow              2000 non-null   float64
 14  snowdepth         1999 non-null   float64
 15  windgust          1746 non-null   float64
 16  windspeed         2000 non-null   float64


# 

# Data Preparation

In [13]:
columns_with_nan = df.columns[df.isnull().any()].tolist()

nan_count_per_column = df[columns_with_nan].isnull().sum()

print("Columns with NaN values and their respective counts:")
print(nan_count_per_column)

Columns with NaN values and their respective counts:
preciptype          1003
snowdepth              1
windgust             254
sealevelpressure       1
severerisk          1374
stations               6
dtype: int64


In [15]:
columns_with_nan = df.columns[df.isnull().any()].tolist()

for column in columns_with_nan:
    unique_values = df[column].unique()
    print(f"Column '{column}':")
    print(unique_values)
    print("---------------------------------")

Column 'preciptype':
['rain' nan 'rain,snow' 'snow']
---------------------------------
Column 'snowdepth':
[0.  0.6 0.4 1.  1.3 3.2 1.4 2.6 2.5 1.1 nan]
---------------------------------
Column 'windgust':
[25.9 14.6 11.2 16.8 36.6 17.2 52.6 38.1 23.5 21.1 16.6 48.2 20.5 16.3
 30.9 61.2 15.4 12.4 30.5 28.1 33.5 26.3 51.8 64.8 22.9 66.6 24.7 55.4
 30.4 30.2 27.8 12.1 24.  11.3 36.  10.5 55.  68.4 21.9 31.3 46.4 25.6
  nan 21.5 13.9 13.1 21.7 16.7 15.9 14.8 57.6 23.  28.4 16.9 14.5 17.8
 18.4 19.9 25.  14.2 29.2 18.5 28.7 31.7 33.9 24.5 29.  13.3 59.4 23.9
 23.2 21.4 50.  24.1 17.  18.6 44.6 22.3 22.5 25.8 19.5 13.7 20.4 22.7
 15.3 16.  17.3 29.5 22.8 14.  14.7 25.5 26.8 26.5 28.8 18.2 21.3 25.7
 17.4 15.8 14.4 23.1 27.4 24.4 19.8 23.8 17.6 19.4 15.7 15.1 18.1 22.6
 22.  37.1 13.2 20.6 25.3 30.7 40.7 20.8 20.2 42.5 38.  18.7 19.6 53.6
 11.8 26.2 29.8 28.3 22.4 43.8 63.  13.  21.2 25.1 18.9 25.4 33.3 12.7
 18.3 28.6 38.3 30.1 13.5 18.  13.8 12.8 35.6 17.5 32.6 28.9 10.6 10.3
 24.3 83.5 48

# 

I looked for the unique values within columns that contain NaN values, to see if missing value is just matter of mistake or it indicates that value for that column was not available in the moment of recording, since it's weather data. 

1. precitype (Precipitation Type) It has 1003 NaN values. Missing values here suggest that the type of precipitation (rain, snow etc.) is not specified or availabe at the moment of recording. It might happen due to not having rain or snow at that moment at all. However, this column doesn't bring much value for my dataset and will drop it. 


2. snowdepth (Snow Depth): There is only 1 NaN value in this column. Snow depth typically represents the depth of accumulated snow at a specific location or time. One missing value indicates that snow depth data was not added as we have 0 within unique values. I will replace NaN value by using Interpolation.


3. windgust (Wind Gust): This column contains 254 missing values. Wind gust refers to the maximum wind speed observed over a brief period. The absence of values in this column for 254 observations signifies that wind gust data is not available for these instances, so I will replace thos NaN values by using Interpolation. 


4. sealevelpressure (Sea Level Pressure): There is 1 missing value in this column. Sea level pressure represents atmospheric pressure adjusted to sea level. One missing value implies that sea level pressure data was simply not added that day. I will replace NaN value using Interpolation. 


5. severerisk (Risk of Severe Weather Conditions): This column has 1374 missing values. Severerisk represents the risk level associated with severe weather conditions. The high count of missing values (1374) indicates that the risk level for severe weather was not applicable for these recordings, therefore NaN values as there was no risk level. In this situation, I will replace NaN values with 0 by using Imputation. 


6. stations: There are 6 missing values in the 'stations' column. This column might contain identifiers or codes for weather stations. The missing values suggest that the station identifiers are not added for these observations. Since this column doesn't bring any value to my dataset, as there are no longitude and altitude for plotting, I will just remove it from dataset. 

# 

### Removing NaN Values

#### Droppping columns

In [16]:
columns_to_drop = ['precip', 'precipprob', 'precipcover', 'preciptype', 'stations']
df.drop(columns=columns_to_drop, inplace=True)

#### Interpolation

In [17]:
df['sealevelpressure'].interpolate(method='linear', inplace=True)
df['windgust'].interpolate(method='linear', inplace=True)
df['snowdepth'].interpolate(method='linear', inplace=True)

#### Imputation

In [18]:
df['severerisk'].fillna(0, inplace=True)

# 

In [19]:
nan_values = df.isnull().sum()
nan_values

datetime            0
tempmax             0
tempmin             0
temp                0
feelslikemax        0
feelslikemin        0
feelslike           0
dew                 0
humidity            0
snow                0
snowdepth           0
windgust            0
windspeed           0
winddir             0
sealevelpressure    0
cloudcover          0
visibility          0
solarradiation      0
solarenergy         0
uvindex             0
severerisk          0
sunrise             0
sunset              0
moonphase           0
conditions          0
description         0
icon                0
dtype: int64